File: scripts/trace-torn-paper.py
#!/usr/bin/env python3
"""
Trace the torn edge from a PNG image and output a high-fidelity SVG path.

The original torn paper PNGs (from Framer) have a colored region with transparency.
We scan each column to find the bottom edge of the opaque region, then map those
coordinates to a 2400×100 SVG viewBox.

Uses Ramer-Douglas-Peucker simplification to keep detail where the edge is complex
while reducing points in smooth areas.

Usage:
    python3 scripts/trace-torn-paper.py public/images/torn-paper-red-1.png "#F92524" > red-path.txt
"""

import sys
import os
import math
from PIL import Image

# Target viewBox
VIEW_W = 2400
VIEW_H = 100

# Alpha threshold
ALPHA_THRESHOLD = 128

# RDP tolerance — lower = more detail, more points
RDP_TOLERANCE = 0.12


def find_bottom_edge(img):
    """For each column, find the bottom-most opaque pixel."""
    width, height = img.size
    pixels = img.load()
    edge = []

    for x in range(width):
        bottom_y = 0
        for y in range(height):
            pixel = pixels[x, y]
            # Handle both RGBA and RGB+palette
            if len(pixel) >= 4:
                alpha = pixel[3]
            else:
                alpha = 255  # fully opaque
            if alpha >= ALPHA_THRESHOLD:
                bottom_y = y
        edge.append((x, bottom_y))

    return edge


def find_top_edge(img):
    """For each column, find the top-most opaque pixel."""
    width, height = img.size
    pixels = img.load()
    edge = []

    for x in range(width):
        top_y = height - 1
        for y in range(height):
            pixel = pixels[x, y]
            if len(pixel) >= 4:
                alpha = pixel[3]
            else:
                alpha = 255
            if alpha >= ALPHA_THRESHOLD:
                top_y = y
                break
        edge.append((x, top_y))

    return edge


def perpendicular_distance(point, line_start, line_end):
    """Distance from point to line defined by line_start and line_end."""
    dx = line_end[0] - line_start[0]
    dy = line_end[1] - line_start[1]
    len_sq = dx * dx + dy * dy

    if len_sq == 0:
        ex = point[0] - line_start[0]
        ey = point[1] - line_start[1]
        return math.sqrt(ex * ex + ey * ey)

    num = abs(dy * point[0] - dx * point[1] +
              line_end[0] * line_start[1] - line_end[1] * line_start[0])
    return num / math.sqrt(len_sq)


def rdp(points, tolerance):
    """Ramer-Douglas-Peucker line simplification."""
    if len(points) <= 2:
        return points

    max_dist = 0
    max_idx = 0
    first = points[0]
    last = points[-1]

    for i in range(1, len(points) - 1):
        d = perpendicular_distance(points[i], first, last)
        if d > max_dist:
            max_dist = d
            max_idx = i

    if max_dist > tolerance:
        left = rdp(points[:max_idx + 1], tolerance)
        right = rdp(points[max_idx:], tolerance)
        return left[:-1] + right
    else:
        return [first, last]


def round_val(n):
    return round(n * 10) / 10


def trace(input_file, fill_color):
    img = Image.open(input_file).convert("RGBA")
    width, height = img.size
    print(f"Image: {width}×{height}", file=sys.stderr)

    # Detect whether this is a top-fill (color at top, torn edge at bottom)
    # or needs special handling. Check a few columns in the middle.
    mid_x = width // 2
    pixels = img.load()

    # Check if top row is opaque (top-fill) or bottom row is opaque (bottom-fill)
    top_alpha = pixels[mid_x, 5][3] if len(pixels[mid_x, 5]) >= 4 else 255
    bot_alpha = pixels[mid_x, height - 5][3] if len(pixels[mid_x, height - 5]) >= 4 else 255

    # Always use bottom edge — the torn paper fills from above,
    # and the torn tear is at the bottom of the opaque region.
    print("Tracing bottom edge of opaque region", file=sys.stderr)
    raw_edge = find_bottom_edge(img)

    print(f"Raw edge points: {len(raw_edge)}", file=sys.stderr)

    # Map to viewBox coordinates
    mapped = [
        (p[0] / width * VIEW_W, p[1] / height * VIEW_H)
        for p in raw_edge
    ]

    # Increase recursion limit for RDP on large point sets
    sys.setrecursionlimit(max(10000, len(mapped) * 2))

    # Simplify with RDP
    simplified = rdp(mapped, RDP_TOLERANCE)
    print(f"Simplified to {len(simplified)} points (tolerance {RDP_TOLERANCE})", file=sys.stderr)

    # If still too many, increase tolerance
    tolerance = RDP_TOLERANCE
    while len(simplified) > 2000:
        tolerance *= 1.5
        simplified = rdp(mapped, tolerance)
        print(f"Re-simplified to {len(simplified)} points (tolerance {tolerance:.3f})", file=sys.stderr)

    # Build SVG path (top-fill: filled from top, torn edge at bottom)
    # Path: M0,0 → H2400 → down to first edge point → trace edge → back to 0,endY → Z
    first = simplified[0]
    last = simplified[-1]

    # The path fills from the top of the viewBox down to the torn edge.
    # We go right-to-left along the edge (since we start at x=2400 after H2400).
    reversed_pts = list(reversed(simplified))

    parts = [f"M0,0 H{VIEW_W} V{round_val(reversed_pts[0][1])}"]

    for pt in reversed_pts[1:]:
        parts.append(f"L{round_val(pt[0])},{round_val(pt[1])}")

    parts.append("Z")
    d = " ".join(parts)

    # Print the d attribute
    print(d)

    # Also write a test SVG file
    out_name = os.path.splitext(os.path.basename(input_file))[0] + "-traced.svg"
    out_dir = os.path.dirname(input_file)
    out_path = os.path.join(out_dir, out_name)

    svg = f"""<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 {VIEW_W} {VIEW_H}" preserveAspectRatio="none" fill="none">
  <path fill="{fill_color}" d="{d}"/>
</svg>"""

    with open(out_path, "w") as f:
        f.write(svg)
    print(f"Wrote test SVG: {out_path}", file=sys.stderr)
    print(f"Path length: {len(d)} chars", file=sys.stderr)


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python3 trace-torn-paper.py <png-file> [fill-color]", file=sys.stderr)
        sys.exit(1)

    input_file = sys.argv[1]
    fill_color = sys.argv[2] if len(sys.argv) > 2 else "#000000"
    trace(input_file, fill_color)

File: scripts/trace-torn-paper.js
#!/usr/bin/env node
/**
 * trace-torn-paper.js
 *
 * Traces the torn edge from a PNG image and outputs a high-fidelity SVG path.
 *
 * The original torn paper PNGs (from Framer) are 4320×600 with transparency.
 * The colored region is opaque; the background is transparent.
 * We scan each column to find the BOTTOM edge of the colored region
 * (last opaque pixel from top), then map those coordinates to a 2400×100 viewBox.
 *
 * Usage:
 *   npm install sharp  (if not installed)
 *   node scripts/trace-torn-paper.js public/images/torn-paper-red-1.png "#F92524"
 *
 * Output: prints an SVG path `d` attribute string to stdout.
 */

const sharp = require("sharp");
const path = require("path");

const [, , inputFile, fillColor = "#000000"] = process.argv;

if (!inputFile) {
  console.error("Usage: node trace-torn-paper.js <png-file> [fill-color]");
  process.exit(1);
}

// Target viewBox dimensions
const VIEW_W = 2400;
const VIEW_H = 100;

// Minimum alpha to consider a pixel "opaque"
const ALPHA_THRESHOLD = 128;

// Adaptive sampling: keep points where the slope changes significantly.
// This reduces point count in smooth areas while preserving detail in jagged areas.
const SLOPE_THRESHOLD = 0.3; // radians — smaller = more points kept
const MAX_GAP = 8; // max viewBox-X units between points (forces points even in smooth areas)
const MIN_GAP = 0.5; // min gap to avoid redundant points

async function trace(file) {
  const img = sharp(path.resolve(file));
  const metadata = await img.metadata();
  const { width, height } = metadata;

  console.error(`Image: ${width}×${height}`);

  // Extract raw RGBA pixel data
  const { data } = await img.raw().ensureAlpha().toBuffer({ resolveWithObject: true });

  // For each column, find the bottom-most opaque pixel (the torn edge)
  // The PNG fills from top; the torn edge is where opacity ends.
  const rawEdge = [];

  for (let x = 0; x < width; x++) {
    let bottomY = 0;
    for (let y = 0; y < height; y++) {
      const idx = (y * width + x) * 4;
      const alpha = data[idx + 3];
      if (alpha >= ALPHA_THRESHOLD) {
        bottomY = y;
      }
    }
    rawEdge.push({ x, y: bottomY });
  }

  console.error(`Raw edge points: ${rawEdge.length}`);

  // Map to viewBox coordinates
  const mapped = rawEdge.map((p) => ({
    x: (p.x / width) * VIEW_W,
    y: (p.y / height) * VIEW_H,
  }));

  // Adaptive simplification using Ramer-Douglas-Peucker algorithm
  const simplified = rdp(mapped, 0.15); // tolerance in viewBox units — lower = more detail

  console.error(`Simplified to ${simplified.length} points`);

  // Ensure we don't exceed a reasonable point count (cap at ~2000)
  // If still too many, increase tolerance
  let finalPoints = simplified;
  if (finalPoints.length > 2000) {
    finalPoints = rdp(mapped, 0.3);
    console.error(`Re-simplified to ${finalPoints.length} points (tolerance 0.3)`);
  }
  if (finalPoints.length > 2000) {
    finalPoints = rdp(mapped, 0.5);
    console.error(`Re-simplified to ${finalPoints.length} points (tolerance 0.5)`);
  }

  // Build SVG path
  // Start: fill from top-left, across top, down to first edge point
  const first = finalPoints[0];
  const last = finalPoints[finalPoints.length - 1];

  let d = `M0,0 H${VIEW_W} V${round(first.y)}`;

  // Trace edge from right to left (our points go left to right, but SVG path
  // should go right to left since we start at top-right after H2400)
  // Actually, let's reverse: start from x=2400, trace leftward
  const reversed = [...finalPoints].reverse();

  for (let i = 0; i < reversed.length; i++) {
    const p = reversed[i];
    d += ` L${round(p.x)},${round(p.y)}`;
  }

  d += " Z";

  // Output
  console.log(d);

  // Also output a complete SVG for testing
  const svg = `<?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 ${VIEW_W} ${VIEW_H}" preserveAspectRatio="none" fill="none">
  <path fill="${fillColor}" d="${d}"/>
</svg>`;

  // Write test SVG
  const outName = path.basename(file, ".png") + "-traced.svg";
  const outPath = path.join(path.dirname(file), outName);
  require("fs").writeFileSync(outPath, svg);
  console.error(`Wrote test SVG: ${outPath}`);
}

/**
 * Ramer-Douglas-Peucker line simplification
 * Removes points that are within `tolerance` distance of the line between their neighbors.
 */
function rdp(points, tolerance) {
  if (points.length <= 2) return points;

  // Find the point with the maximum distance from the line between first and last
  let maxDist = 0;
  let maxIdx = 0;

  const first = points[0];
  const last = points[points.length - 1];

  for (let i = 1; i < points.length - 1; i++) {
    const d = perpendicularDistance(points[i], first, last);
    if (d > maxDist) {
      maxDist = d;
      maxIdx = i;
    }
  }

  if (maxDist > tolerance) {
    // Recurse on both halves
    const left = rdp(points.slice(0, maxIdx + 1), tolerance);
    const right = rdp(points.slice(maxIdx), tolerance);
    return [...left.slice(0, -1), ...right];
  } else {
    // All points between first and last are within tolerance — discard them
    return [first, last];
  }
}

function perpendicularDistance(point, lineStart, lineEnd) {
  const dx = lineEnd.x - lineStart.x;
  const dy = lineEnd.y - lineStart.y;
  const lenSq = dx * dx + dy * dy;

  if (lenSq === 0) {
    // lineStart and lineEnd are the same point
    const ex = point.x - lineStart.x;
    const ey = point.y - lineStart.y;
    return Math.sqrt(ex * ex + ey * ey);
  }

  const num = Math.abs(dy * point.x - dx * point.y + lineEnd.x * lineStart.y - lineEnd.y * lineStart.x);
  return num / Math.sqrt(lenSq);
}

function round(n) {
  return Math.round(n * 10) / 10;
}

trace(inputFile).catch((err) => {
  console.error(err);
  process.exit(1);
});
